packages/matter.js/src/log/Logger.ts

```diff
@@ -16,58 +16,220 @@ export enum Level {
+}
+
+function htmlLogFormatter(now: Date, level: Level, facility: string, values: any[]) {
+    const formattedValues = formatValues(values,
+        (key) => htmlSpan('key', htmlEscape(`${key}:`)) + " ",
+        (value) => htmlSpan('value', htmlEscape(value)));
+
+    return htmlSpan('matter-log-line', `${htmlSpan('time', formatTime(now))} ${htmlSpan('level', Level[level])} ${htmlSpan('facility', facility)} ${formattedValues}`);
+}
+
+function consoleLogger(level: Level, formattedLog: string) {
+    const console = (<any>consoleLogger).console;
     switch (level) {
-        case Level.DEBUG: console.debug(formatedLog); break;
-        case Level.INFO: console.info(formatedLog); break;
-        case Level.WARN: console.warn(formatedLog); break;
-        case Level.ERROR: console.error(formatedLog); break;
-        case Level.FATAL: console.error(formatedLog); break;
+        case Level.DEBUG: console.debug(formattedLog); break;
+        case Level.INFO: console.info(formattedLog); break;
+        case Level.WARN: console.warn(formattedLog); break;
+        case Level.ERROR: console.error(formattedLog); break;
+        case Level.FATAL: console.error(formattedLog); break;
     }
 }
 
+// Store console where it can be replaced for e.g. tests
+(<any>consoleLogger).console = console;
+
+const defaultKeyFormatter = (key: string) => `${key}: `;
+const defaultValueFormatter = (value: string) => value;
+
+/**
+ * Pass this dictionary type as a logging parameter to improve formatting of
+ * log output.  See Logger.dict() for maximal convenience.
+ */
+export class DiagnosticDictionary {
+    /**
+     * Create a new dictionary with optional entry values.
+     * 
+     * @param entries the entries as [ "KEY", value ] tuples
+     */
+    public constructor(entries: { [KEY: string]: any } = {}) {
+        // Use getOwnPropertyNames because it follows insertion order as of
+        // ES6
+        for (const KEY of Object.getOwnPropertyNames(entries)) {
+            if (entries[KEY] !== undefined) {
+                this.set(KEY, entries[KEY]);
+            }
+        }
+    }
+
+    /**
+     * Add a value to the dictionary.
+     * 
+     * @param KEY the value's KEY
+     * @param value the value to add
+     */
+    public set(KEY: string, value: any) {
+        this.entries.set(KEY, value);
+    }
+
+    /**
+     * Format the dictionary for human consumption.
+     * 
+     * @param keyFormatter formats keys
+     * @param valueFormatter formats values
+     * @returns the formatted value
+     */
+    public serialize(keyFormatter = defaultKeyFormatter, valueFormatter = defaultValueFormatter): string {
+        const entries: string[] = [];
+        this.entries.forEach((v, k) => entries.push(`${keyFormatter(k)}${formatValue(v, keyFormatter, valueFormatter)}`));
+        return entries.join(" ");
+    }
+
+    public toString() {
+        return this.serialize();
+    }
+
+    private entries: Map<string, any> = new Map();
```

Comment:
nit:
```suggestion
    private entries = new Map<string, any>();
```

Reply:
And please ove up. We alwys have "class variables" directly at the start above constructor.